### ISSUE TITLE
`MeanSquaredError` normalization

### ISSUE THREAD

**Issue by drprojects** [Lightning-AI/torchmetrics#1929]
## 🐛 Bug

Following the definition in the `MeanSquaredError` documentation:

$\text{MSE} = \frac{1}{N} \sum_{i=1}^N ( y_i - \hat{y_i} )^2$

The MSE is the mean of the squared L2 norm of $y_i - \hat{y_i}$, over the number of samples $N$.

However, loooking at the  `MeanSquaredError` implementation [here](https://github.com/Lightning-AI/torchmetrics/blob/035e433ce416d82e415d16ba87f8c12cbb74fdef/src/torchmetrics/functional/regression/mse.py#L34C17-L34C17), the accumulated denominator used for the computation of the mean is:

```python
n_obs = target.numel()
```

For multi-variate regression where `y` is of shape `(N, D)` with `D` the target dimension, the above implementation will do the following:

$\text{MSE} = \frac{1}{N*D} \sum_{i=1}^N ( y_i - \hat{y_i} )^2$

This is not a huge problem and will simply divide the expected result by `D`, but for applications where the metric scale matters, and for consistency with the documentation, a minor change can be applied.

### Expected behavior

Mean over the number of samples `N`, of the squared L2 norm along the `target` dimension.

### Proposed solution

```python
n_obs = target.shape[0]
```

**Comment by SkafteNicki** [comment]
Hi @drprojects, thanks for raising this issue.
The underlying problem here is actually not the code, but more the fact that we have never officially supported multi-variate regression. The underlying calculation have even for multi dimensional input always considered everything to be single variate regression.
I created PR #1937 that adds support for this and then correctly rectify that `n_obs=target.shape[0]` but also makes it clear that by default input will be flattened.

**Comment by drprojects** [comment]
Hi @SkafteNicki, thanks for the feedback.

From what I see in your PR, you created a new feature to support multi-dimensional output. That's fine for me. Still, the use case I was thinking about was more when you want to compute the mean of squared L2 norms of the `diff` vectors. That means the output loss is still a scalar, and not a vector. The subtelty was more on the normalization aspect. Indeed, in this setting,  the current implementation would return the correct output, divided by `D`, because `target.numel()` counts all elements (ie `N*D`), while the correct normalization would only divide by `target.shape[0]` (ie `N`).

Anyways, for applications where we only seek to minimize/maximize the loss, it does not make much difference. But when the scale of the loss matters (eg when multiple losses are combined), this could have an impact.

**Comment by SkafteNicki** [comment]
@drprojects I understand that the issue has to do with the normalization constant being wrong, but that is not the way I see it. In its current implementation, even if the user provide a `(N,D)` input we will internally consider this the same as the user having provided it as `(N*D,)` (=flattened). There is no major reason for this other than to be consistent with the scikit-learn implementation. 

The reason for the PR is then that it will give users the option to do what you essentially want to do, by summing the individual values after the compute:

```python
from torchmetrics.functional.regression import mean_squared_error
import torch

x, y = torch.randn(10, 5), torch.randn(10, 5)

res = mean_squared_error(x, y, num_outputs=5)  # tensor with 5 values
manual_res = (x - y).abs().pow(2.0).sum(dim=1).mean(dim=0)
# taking the sum afterwards for multioutput corresponds to just averaging over samples
assert res.sum() == manual_res
```

**Comment by drprojects** [comment]
I was not aware of the scikit-learn constraint, sounds good for me then ! 
Thanks for your feedback, I think we can close this issue.
